### ISSUE TITLE
AzureRM backend support for users to plan without permission to apply

### ISSUE THREAD

**Issue by aidapsibr** [hashicorp/terraform#28423]
### Current Terraform Version

```
0.15.0
```

### Use-cases

Enable users to access a particular storage container with only `Storage Blob Data Reader` for performing no-lock plans to validate changes locally.

### Attempted Solutions

Attempted to assign the role and leverage the ` use_azuread_auth` option on the backend to access a particular storage container.

### Proposal

Currently, authorization fails even with ` use_azuread_auth` and appropriate data roles, because the backend tries to listKeys on the storage account. Updating the backend flow to skip this step and instead use AzureAD tokens to access the blobs would enable this directly. This should be a non-breaking change I believe since storage account owner (listKeys) has the blob data access roles.

### References

#20831 - The title of this was misleading, you no longer *directly* need to provide a key, but the backend will still use the keys to access blobs.

**Comment by aidapsibr** [comment]
This work would also enable `Storage Blob Data Owner` to be sufficient for running apply and locked plans as an intended side-effect.

**Comment by JustinGrote** [comment]
@aidapsibr Curious why this has been backlogged so long. The code just has to be adjusted to use RBAC for the operations rather than an account key, and the way it currently exists means each state file must have its own storage account for security since you basically use the "root password" for all operations, making RBAC like OIDC meaningless and misleading in the documentation.

Users who do not know this esoteric detail are going to think they are safe by putting their state files in separate containers within a storage account, whereas anyone who compromises the terraform access (or simply makes a mistake) can delete *all* the files in *all* containers, meaning the annoying overhead of one storage account per state file is necessary for proper security and isolation.

**Comment by aidapsibr** [comment]
I'm just the one who opened the issue, our solution was to do the same as you described with separate storage accounts which is really unfortunate. Wish this would be prioritized.

**Comment by yann-soubeyrand** [comment]
Isn’t this issue fixed by using `use_azuread_auth = true` in Terraform azurerm backend configuration block?

**Comment by BzSpi** [comment]
I don't think so since user must be Data Owner to have the lock permission.

**Comment by KoenR3** [comment]
This is still a major security issue. If yoy disable access keys on a storage account, remote state no longer functions. The ability to remove the use of root keys is critical in moving everything to mamaged identities.

**Comment by JustinGrote** [comment]
The latest versions seem to do this correctly, I've been able to not provide an access key and have everything in TFState work with RBAC, so my tfstate for different environments can be two containers in the same storage account and work fine now.

**Comment by KoenR3** [comment]
Yes you do not have to pro ide an access key but when turning off access keys on the storage account the remote state fails...

**Comment by MarkKharitonov** [comment]
What is the status of it? We are facing the same issue.

**Comment by nrjohnstone** [comment]
This seems to work with normal user accounts, but once you use managed identities it goes back to needing permission to list the storage keys, even if the managed identity has the correct permissions to the blob storage container itself

**Comment by g13013** [comment]
Any news about this ?

**Comment by g13013** [comment]
This is a security issue! why is this not prioritized ?

**Comment by aidapsibr** [comment]
Does anyone know if OpenTofu has fixed this problem? If not someone should mirror this issue. I'm not longer using Azure, so not the best representative for it. I imagine this will have more traction there. I've given up on hashicorp, I was a paying customer of terraform cloud for the years this was open too...

**Comment by rcskosir** [comment]
Thanks for taking the time to submit this issue. It looks like this has been resolved in more recent versions of Terraform (1.2.0 and 1.4.6 like mentioned above). As such, I am going to mark this issue as closed. If that is not the case, please provide additional information including the recent version in which you are not seeing this enhancement, thanks!
